### PR TITLE
Task hooks

### LINKED ISSUE #76: Task completed callbacks
Once a task is completed, we should be able to perform actions based on the given result.

image

### PR DESCRIPTION
**Pull request by ismellike**
Closes #76 

Adds a new `mock-hook-consumer` contract to test out all the of the different task hooks.

Adds a test in the `task-queue` contract to use the `mock-hook-consumer`.
- The test here is using the open payment config to allow the `mock-hook-consumer` to create a new task when a task is completed
- When a task is created, the `mock-hook-consumer` is simply incrementing its created counter.
- When a task is timed out, the `mock-hook-consumer` will error out, but the error will be handled correctly by the `task-queue` to prevent blocking.
- When a task is completed, the `mock-hook-consumer` will create a new task based on the given task's result. (square the square)

### CODE REVIEW

File: contracts/task-queue/src/tests/common.rs [hunk 1]
```diff
@@ -455,6 +466,69 @@ where
...
+    let counter = mock_consumer.created_count().unwrap();
+    assert!(counter.is_zero());
+
+    // Create a task
+    let payload = json!({"x": 5});
+    let task_id = make_task_with_funds(&task_contract, "Test Task", None, &payload, &funds);
+
+    // Verify task created hook
+    let counter = mock_consumer.created_count().unwrap();
+    assert_eq!(counter, Uint128::one());
```
Comment (ueco-jb) [inline contracts/task-queue/src/tests/common.rs:504]: Nitpicking, but seems like a regular `u128` for a counter would be easier to use and read.

File: contracts/mock-hook-consumer/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+            contract_addr: task_queue.to_string(),
+            msg: to_json_binary(&lavs_apis::tasks::ExecuteMsg::Custom(
+                lavs_apis::tasks::CustomExecuteMsg::Create {
+                    description: task.description,
+                    timeout: None,
+                    payload: serde_json::to_value(request).map_err(|e| {
+                        StdError::generic_err(format!(
+                            "Could not serialize request into JSON: {}",
+                            e
+                        ))
```
Comment (ueco-jb) [inline contracts/mock-hook-consumer/src/contract.rs:153]: This is one thing that puzzles me, I assume that this is just an example of an operation handled automatically after the one from the queue is finished?

**Review by ueco-jb — APPROVED** [review]
Good and clean example of how to implement hooks.
Depending on the level of devs who are supposed to read from that example, you might want to explicitely mention in the readme that the cw-minus hook's interface is being used here (if that's not obvious :) ).

File: contracts/mock-hook-consumer/src/contract.rs (hunk 2 above)
Comment (ismellike) [reply contracts/mock-hook-consumer/src/contract.rs:153]: Yeah, this was an example of a task completion triggering the creation of another task: https://github.com/Lay3rLabs/avs-toolkit/issues/76#issuecomment-2397918372

Comment (ueco-jb) [reply contracts/mock-hook-consumer/src/contract.rs:153]: I see, thank you :+1:

**Review by ethanfrey — APPROVED** [review]
Looks good. Going to merge as discussed in #76 as it works and a longer-term design is a bit furtehr off.

I notice there is no update the the cli tool to actually add these hooks. It would be good to add in a follow up.

**Pull request merged**